### PR TITLE
7.0 refactor account_partner_required

### PR DESCRIPTION
**Pull request by sbidoul**
refactor account_partner_required:
- add extension points for policies
- use constraints for readbility and robustness
- slightly adapt the test to work around an issue in account_constraints (which allows the simplification of the travis configuration)

### CODE REVIEW

**Comment by sbidoul** [comment]
Tests ok. Travis red because of pylint errors in unrelated modules, fixed in #47

**Comment by sbidoul** [comment]
Help! I don't understand why Travis complains.

**Comment by sbidoul** [comment]
Okay, got it, it's the missing ACL in account_credit_control_dunning_fees. maintainer-quality-tools move too fast. Sigh.

**Comment by sbidoul** [comment]
@nbessi do you feel like adding the missing ACL in credit.control.dunning.fees.computer? Can't that be a TransientModel?

**Comment by nbessi** [comment]
@sbidoul what ??

**Comment by sbidoul** [comment]
@nbessi I tried to explain in more details in #49

**Comment by laetitia-gangloff** [comment]
:+1:  code and test ok for me

File: account_partner_required/account.py [hunk 1]
```diff
@@ -27,12 +27,21 @@
...
+    def __get_policies(self, cr, uid, context=None):
+        """ Call method which can be inherited """
+        return self._get_policies(cr, uid, context=context)
+
     _columns = {
         'partner_policy': fields.selection(
-            [('optional', 'Optional'),
-             ('always', 'Always'),
-             ('never', 'Never')],
+            __get_policies,
```
Comment (pedrobaeza) [inline account_partner_required/account.py]: @bwrsandman gave me a hint to get this called without this trick using a lambda expression. Can you refresh us the method, Sandy?

Comment (sbidoul) [inline account_partner_required/account.py]: @pedrobaeza I suppose I could replace __get_policies by a lambda that calls _get_policies, but the net result is the same.

Comment (pedrobaeza) [inline account_partner_required/account.py]: They say that you don't need the double method. Inheritance is caught properly. I didn't try.

**Comment by sbidoul** [comment]
Rebased. All green thanks to @nbessi's work on #49 .

File: account_partner_required/account.py (hunk 1 above)
Comment (bwrsandman) [inline account_partner_required/account.py]: This is setting the `id(account_account_type.__get_policies)` as the function from this class, which means you cannot override it.

Using a lambda `self, *a, **kw: self.__get_policies(*a, **kw)` sets the function as the `id` of the lambda which calls on `self` which can be overridden.

In 8.0, you don't need this when you pass the function name as a string.

File: account_partner_required/account.py [hunk 2]
```diff
@@ -27,12 +27,21 @@
 class account_account_type(orm.Model):
     _inherit = "account.account.type"
 
+    def _get_policies(self, cr, uid, context=None):
+        """This is the method to be inherited for adding policies"""
+        return [('optional', 'Optional'),
+                ('always', 'Always'),
+                ('never', 'Never')]
```
Comment (bwrsandman) [inline account_partner_required/account.py]: The second element of these tuples need to be translated using `_('xxxxx')`

Comment (sbidoul) [inline account_partner_required/account.py]: Good catch, thanks.

**Comment by sbidoul** [comment]
@pedrobaeza @bwrsandman I applied the lambda thing. It does not change the behaviour but it's more readable. Thanks for the review.

**Comment by bwrsandman** [comment]
:+1:

**Comment by pedrobaeza** [comment]
Thanks :+1:

**Pull request merged**